pkg/ucp/store/postgres/postgresclient.go

```diff
@@ -0,0 +1,422 @@
+	"context"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/jackc/pgx/v5"
+	"github.com/jackc/pgx/v5/pgconn"
+	"github.com/radius-project/radius/pkg/to"
+	"github.com/radius-project/radius/pkg/ucp/resources"
+	"github.com/radius-project/radius/pkg/ucp/store"
+	"github.com/radius-project/radius/pkg/ucp/store/storeutil"
+	"github.com/radius-project/radius/pkg/ucp/util/etag"
+)
+
+// PostgresAPI defines the API surface from pgx that we use. This is used to allow for easier testing.
+//
+// Keep these definitions in sync with pgxpool.Pool and pgx.Conn.
+type PostgresAPI interface {
+	// Exec executes a query without returning any rows.
+	Exec(ctx context.Context, sql string, args ...any) (pgconn.CommandTag, error)
+	// QueryRow executes a query that is expected to return at most one row.
+	QueryRow(ctx context.Context, sql string, args ...any) pgx.Row
+	// Query executes a query that returns rows.
+	Query(ctx context.Context, sql string, args ...any) (pgx.Rows, error)
+}
+
+// NewPostgresClient creates a new PostgresClient.
+func NewPostgresClient(api PostgresAPI) *PostgresClient {
+	return &PostgresClient{api: api}
+}
+
+var _ store.StorageClient = (*PostgresClient)(nil)
+
+// PostgresClient is a storage client that uses Postgres as the backend.
+type PostgresClient struct {
+	api PostgresAPI
+}
+
+// Delete implements store.StorageClient.
+func (p *PostgresClient) Delete(ctx context.Context, id string, options ...store.DeleteOptions) error {
+	if ctx == nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'ctx' is required"}
+	}
+
+	parsed, err := resources.Parse(id)
+	if err != nil {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must be a valid resource id"}
+	}
+	if parsed.IsEmpty() {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must not be empty"}
+	}
+	if parsed.IsResourceCollection() || parsed.IsScopeCollection() {
+		return &store.ErrInvalid{Message: "invalid argument. 'id' must refer to a named resource, not a collection"}
+	}
+
+	converted, err := storeutil.ConvertScopeIDToResourceID(parsed)
+	if err != nil {
+		return err
+	}
+
+	config := store.NewDeleteConfig(options...)
+	var etag *string
+	if config.ETag != "" {
+		etag = &config.ETag
+	}
+
+	// We need different SQL for the case where an etag is provided vs not provided.
+	//
+	// The key behavior difference is that if an etag is provided, should report failure differently.
+	sql := `
+WITH deleted AS (
+	DELETE FROM resources
+	WHERE id = $1
+	RETURNING id
+)
+SELECT
+CASE
+	WHEN EXISTS (SELECT 1 FROM deleted) THEN 'Success'
+	WHEN EXISTS (SELECT 1 FROM resources WHERE id = $1) THEN 'ErrConcurrency'
+	ELSE 'ErrNotFound'
+END AS result;`
+
+	args := []any{storeutil.NormalizePart(converted.String())}
+
+	if config.ETag != "" {
+		// NOTE: we want to report ErrConcurrency for all failure cases here. This is what the tests do.
+		sql = `
+WITH deleted AS (
+	DELETE FROM resources
+	WHERE id = $1 AND etag = $2
```

Comment:
We can make the **sql** variable an **fmt.Sprintf** and just update the WHERE part of the query instead of writing two almost identical queries.

Reply:
I see that they also return two different results. Ignore my comment above.

Reply:
I'm really averse to dynamically generating SQL. So not doing it until it's an absolute requirement.